### ISSUE TITLE
Some questions about the training.

### ISSUE THREAD

**Issue by innocentlzx** [Choddeok/EmoSpherepp#11]
Thanks for sharing the code!
When I run with the .ckpt you share, I can get the right results, but when I train it myself, I found many errors in the code, like key error "mel_losses", I add these missing keys in emospherepp.yaml, but when I finish the training, the results are all the sound that can not be understand, here are my final configs, should I modify some of it? Thanks!
CE_loss: false
accumulate_grad_batches: 1
act_fn: snakebeta
amp: false
attention_head_dim: 64
audio_num_mel_bins: 80
audio_sample_rate: 16000
base_config:
- ./base.yaml
beta_max: 20.0
beta_min: 0.05
binarization_args:
  min_sil_duration: 0.1
  shuffle: false
  test_range:
  - 0
  - 100
  train_range:
  - 200
  - -1
  trim_eos_bos: false
  valid_range:
  - 100
  - 200
  with_align: true
  with_f0: true
  with_f0cwt: false
  with_linear: false
  with_spk_embed: false
  with_wav: false
binarizer_cls: data_gen.tts.base_binarizer.BaseBinarizer
binary_data_dir: /workspace/choddeok/hd0/dataset/binary/Fianl_ESD_VAD2ESV_I2I_IQR
channels:
- 256
- 256
check_val_every_n_epoch: 10
clip_grad_norm: 1
clip_grad_value: 0
dataset_cls: tasks.tts.dataset_utils.EmoSpherepp_Dataset
debug: false
dec_dim: 64
disc_interval: 1
disc_lr: 0.0001
disc_norm: in
disc_reduction: stack
disc_start_steps: 0
disc_win_num: 3
discriminator_optimizer_params:
  eps: 1.0e-06
  weight_decay: 0.0
discriminator_scheduler_params:
  gamma: 0.5
  step_size: 40000
discriminator_train_start_steps: 0
down_block_type: transformer
dropout: 0.05
ds_workers: 1
dur_predictor_kernel: 3
dur_predictor_layers: 2
emo_inten: null
emo_norm: ''
emo_style: null
enc_dropout: 0.1
enc_kernel: 3
endless_ds: true
eval_max_batches: -1
f0_max: 800
f0_min: 80
fft_size: 1024
filter_channels: 768
filter_channels_dp: 256
fmax: 8000
fmin: 0
for_extractor: true
frames_multiple: 1
gen_dir_name: p_seen
griffin_lim_iters: 30
hop_size: 256
infer: false
intensity: ''
jcu_hidden: 256
lambda_mel_adv: 0.05
load_ckpt: ''
loud_norm: false
lr: 0.0001
lr_d: 0.0001
max_epochs: 1000
max_frames: 1548
max_input_tokens: 1550
max_sentences: 32
max_tokens: 40000
max_updates: 1100000
max_valid_sentences: 1
max_valid_tokens: 60000
max_wav_value: 32768.0
mel_disc_hidden_size: 128
mel_lengths:
- 48
- 64
- 96
mel_losses: l1:0.5|ssim:0.5
mel_vmax: 2.1758
mel_vmin: -9.0314
mid_block_type: transformer
min_frames: 0
model_cls: models.tts.EmoSpherepp.EmoSpherepp
n_blocks: 1
n_enc_channels: 192
n_enc_layers: 6
n_heads: 2
n_spks: 10
noise_scale: 0.8
num_ckpt_keep: 3
num_emo: 5
num_heads: 2
num_mid_blocks: 2
num_sanity_val_steps: 5
num_spk: 10
num_valid_plots: 10
optimizer_adam_beta1: 0.9
optimizer_adam_beta1_d: 0.5
optimizer_adam_beta2: 0.98
optimizer_adam_beta2_d: 0.9
out_dims: 80
out_size: 64
out_wav_norm: false
pe_scale: 1000
pitch_extractor: parselmouth
pitch_key: pitch
post_glow_hidden: 128
post_glow_kernel_size: 3
post_glow_n_block_layers: 3
post_glow_n_blocks: 8
post_share_cond_layers: false
predictor_dropout: 0.5
predictor_kernel: 5
predictor_layers: 5
preprocess_args:
  add_eos_bos: true
  mfa_group_shuffle: false
  mfa_offset: 0.02
  nsample_per_mfa_group: 1000
  reset_phone_dict: true
  reset_word_dict: true
  save_sil_mask: true
  txt_processor: en
  use_mfa: true
  vad_max_silence_length: 12
  wav_processors: []
  with_phsep: true
preprocess_cls: ''
print_nan_grads: false
processed_data_dir: /workspace/choddeok/hd0/dataset/processed/ESD_all
profile_infer: false
raw_data_dir: ''
rename_tmux: true
resume_from_checkpoint: 0
save_best: false
save_codes:
- tasks
- modules
- egs
save_f0: false
save_gt: false
save_mel_npy: true
save_plot_png: false
scheduler: warmup
seed: 1234
share_wn_layers: 4
sigma_min: 1e-4
sigmoid_scale: false
solver: euler
sort_by_len: true
spk_emb: ''
spk_emb_dim: 64
sty_lambda_reversal: 0.1
style: ''
style_norm: ''
task_cls: tasks.tts.EmoSpherepp.EmoSpherepp
tb_log_interval: 100
test_ids: []
test_input_yaml: ''
test_num: 100
test_set_name: test
train_set_name: train
train_sets: ''
up_block_type: transformer
use_emo_embed: false
use_emo_id: true
use_emo_lookup: true
use_latent_cond: true
use_mlp_style: ''
use_pitch_embed: false
use_spk_embed: false
use_spk_id: true
use_spk_lookup: true
use_txt_cond: true
use_word_input: false
val_check_interval: 2000
valid_infer_interval: 20000
valid_monitor_key: val_loss
valid_monitor_mode: min
valid_plots:
- 0
- 1
- 20
- 21
- 40
- 41
- 60
- 61
- 80
- 81
valid_set_name: valid
vocoder: BigVGAN
vocoder_ckpt: /workspace/choddeok/hd0/checkpoints/vocoder/bigV_16k/g_04500000
vocoder_cls: models.vocoder.bigvgan.BigVGAN
vocoder_config: configs/models/vocoder/bigvgan_16k.yaml
warmup_updates: 4000
weight_decay: 0
win_size: 1024
window_size: 4
word_dict_size: 10000
work_dir: /workspace/choddeok/hd0/out/NAT_checkpoints/EmoSpherepp

**Comment by nebulajo** [comment]
Have you resolved this issue? I'm experiencing the same problem.

**Comment by innocentlzx** [comment]
> Have you resolved this issue? I'm experiencing the same problem.

Sorry，I still have no idea how to solve it, the code has too many errors.

**Comment by Hanzhiying970** [comment]
> Have you resolved this issue? I'm experiencing the same problem.你解决这个问题了吗？我也遇到了同样的问题。

hello,Have you resolved this issue? I'm experiencing the same problem.

**Comment by Hanzhiying970** [comment]
hello,can you share me bigV_16k/g_04500000
and bigvgan_16k.yaml? thank you! 2093046865@qq.com

**Comment by Choddeok** [comment]
@innocentlzx @nebulajo @Hanzhiying970 

Hi, sorry for the late reply.

Regarding the mel_losses key error, do you mean that the mel_losses variable is not defined during training?

From what you described, I think the most likely issue is related to the YAML loading path and the directory paths specified in the config files. In the config you sent, there are still several path settings such as:

binary_data_dir: /workspace/choddeok/hd0/dataset/binary/Fianl_ESD_VAD2ESV_I2I_IQR
processed_data_dir: /workspace/choddeok/hd0/dataset/processed/ESD_all
vocoder_ckpt: /workspace/choddeok/hd0/checkpoints/vocoder/bigV_16k/g_04500000
work_dir: /workspace/choddeok/hd0/out/NAT_checkpoints/EmoSpherepp

If these paths are different in your environment, the model may not be loading the correct data, vocoder checkpoint, or training outputs properly, which could lead to unintelligible synthesized audio even if training finishes.

**Comment by Choddeok** [comment]
Also, if the overall code framework feels a bit difficult to follow, it may be easier to run the code by referring to the NATSpeech framework that this project is based on:
https://github.com/NATSpeech/NATSpeech

I think that would help you understand the overall structure and execution flow more easily.

I also set it up so that TensorBoard can be used, so I would recommend monitoring the intermediate results there as well to check whether understandable audio is being generated during training.

If you can share a bit more detail about the issue you are seeing, I’ll do my best to help.

**Comment by Hanzhiying970** [comment]
> Also, if the overall code framework feels a bit difficult to follow, it may be easier to run the code by referring to the NATSpeech framework that this project is based on:另外，如果整体代码框架感觉有点难以理解，运行代码可能更简单，参考本项目所基于的 NATSpeech 框架： https://github.com/NATSpeech/NATSpeech
> 
> I think that would help you understand the overall structure and execution flow more easily.我觉得这会帮助你更容易理解整体结构和执行流程。
> 
> I also set it up so that TensorBoard can be used, so I would recommend monitoring the intermediate results there as well to check whether understandable audio is being generated during training.我还设置了可以使用 TensorBoard，所以我建议你也监控中间结果，检查训练过程中是否生成了可理解的音频。
> 
> If you can share a bit more detail about the issue you are seeing, I’ll do my best to help.如果你能详细说明你遇到的问题，我会尽力帮忙。

/home/hanzhiying/atest/EmoSpherepp/egs/egs_bases/tts/emospherepp.yaml：
base_config: ./base.yaml
task_cls: ''
dataset_cls: ''
model_cls: ''

n_enc_channels: 192
filter_channels: 768
filter_channels_dp: 256
n_heads: 2
n_enc_layers: 6
enc_kernel: 3
enc_dropout: 0.1
window_size: 4
dec_dim: 64
beta_min: 0.05
beta_max: 20.0
pe_scale: 1000

use_spk_embed: false
use_spk_id: true
n_spks: 10
spk_emb_dim: 64

solver: euler
sigma_min: 1e-4

channels: [256, 256]
dropout: 0.05
attention_head_dim: 64
n_blocks: 1
num_mid_blocks: 2
num_heads: 2
act_fn: snakebeta
down_block_type: "transformer"
mid_block_type: "transformer"
up_block_type: "transformer"

out_size: 64

lr: 0.0001
scheduler: warmup # rsqrt|warmup|none|noam
# train and eval
max_sentences: 16 #32
warmup_updates: 4000

max_updates: 1100000
valid_infer_interval: 20000

intensity: ""
style: ""
spk_emb: ""

# duration
dur_predictor_kernel: 3
dur_predictor_layers: 2
predictor_kernel: 5
predictor_layers: 5
predictor_dropout: 0.5


# 新增配置，放在任意位置（如duration配置下方）
pitch_type: frame



#新增参数
mel_losses: "l1:1.0"
out_dims: 80
# 声码器相关配置
#vocoder: HifiGan
#vocoder_cls: models.vocoder.hifigan.HiFiGAN
#vocoder_config: /home/hanzhiying/atest/EmoSpherepp/pretrained/checkpoints/vocoder/hifigan_universal_v1/config.json  # 替换为你实际的配置文件路径
#vocoder_ckpt: /home/hanzhiying/atest/EmoSpherepp/pretrained/checkpoints/vocoder/hifigan_universal_v1/g_02500000
vocoder: BigVGAN
vocoder_cls: models.vocoder.bigvgan.BigVGAN
vocoder_config: /home/hanzhiying/atest/EmoSpherepp/pretrained/checkpoints/vocoder/bigV_16k/config.json
vocoder_ckpt: '/home/hanzhiying/atest/EmoSpherepp/pretrained/checkpoints/vocoder/bigV_16k/g_00500000'


Hello, sorry to bother you. I modified the last few lines of parameters in the configuration file according to the errors reported during training, and I’m not sure if these changes are correct. The audio generated using this configuration sounds very robotic and unclear. I would like to ask if this is caused by not training a dedicated 16k vocoder, or if there are issues with my parameter settings? I would appreciate it if you could help clarify this. In addition, could you please share the code for the visualization part in the paper? Thank you very much for your time and help.

**Comment by Choddeok** [comment]
@Hanzhiying970 

Hello,

Thank you for your question. Based on the issue you described, it is more likely related to **hyperparameter settings** or a **mismatch in the vocoder loading/configuration**, rather than a fundamental problem with the training itself.

First, I would recommend checking whether the **mel-spectrogram settings used for the vocoder and the TTS model are exactly the same**. From what I understand, the configuration used in BigVGAN may differ, so this mismatch could lead to degraded audio quality.

For reference, you may want to verify parameters such as:

```yaml
audio_sample_rate: 
hop_size: 
win_size: 
fft_size: 
fmin: 
fmax: 
f0_min: 
f0_max: 
```

In particular, please double-check the following:

1. Whether the **mel extraction settings** are fully aligned between the vocoder and the TTS training setup
2. Whether the **vocoder loading code** is correctly configured

Additionally, could you confirm if you modified the part where the vocoder is loaded from
`EmoSpherepp/tasks/vocoder/bigvgan.py`
to use the `hifigan.py` file that I uploaded?

To further diagnose the issue, I recommend first listening to the **vocoded audio (output directly from the vocoder)** before full TTS inference. If the audio already sounds unnatural or distorted at this stage, the issue is likely with the **vocoder configuration or loading**, rather than the training process.

Therefore, the robotic and unclear audio you are experiencing is not necessarily caused by the absence of a dedicated 16k vocoder. It is more likely due to a **configuration mismatch or incorrect vocoder setup**.

Thank you for your time, and please let me know if you need further clarification.

**Comment by Hanzhiying970** [comment]
> [@Hanzhiying970](https://github.com/Hanzhiying970)
> 
> Hello,  你好，
> 
> Thank you for your question. Based on the issue you described, it is more likely related to **hyperparameter settings** or a **mismatch in the vocoder loading/configuration**, rather than a fundamental problem with the training itself.谢谢你的提问。根据你描述的问题，更可能是**超参数设置**或**声码器加载/配置不匹配 **，而不是训练本身的根本问题。
> 
> First, I would recommend checking whether the **mel-spectrogram settings used for the vocoder and the TTS model are exactly the same**. From what I understand, the configuration used in BigVGAN may differ, so this mismatch could lead to degraded audio quality.首先，我建议检查声**码器和 TTS 模型所用的 mel-spectrogram 设置是否完全相同 **。据我了解，BigVGAN 所用配置可能不同，这种不匹配可能导致音质下降。
> 
> For reference, you may want to verify parameters such as:作为参考，你可能需要验证以下参数：
> 
> audio_sample_rate: 
> hop_size: 
> win_size: 
> fft_size: 
> fmin: 
> fmax: 
> f0_min: 
> f0_max: 
> In particular, please double-check the following:特别是，请仔细核对以下内容：
> 
> 1. Whether the **mel extraction settings** are fully aligned between the vocoder and the TTS training setupmel 提取**设置**是否完全对齐了声码器和 TTS 训练设置
> 2. Whether the **vocoder loading code** is correctly configured声**码器加载代码**是否配置正确
> 
> Additionally, could you confirm if you modified the part where the vocoder is loaded from另外，你能确认你是否修改了声码器加载的那个部分吗 `EmoSpherepp/tasks/vocoder/bigvgan.py` to use the `hifigan.py` file that I uploaded?用我上传的 `hifigan.py` 文件？
> 
> To further diagnose the issue, I recommend first listening to the **vocoded audio (output directly from the vocoder)** before full TTS inference. If the audio already sounds unnatural or distorted at this stage, the issue is likely with the **vocoder configuration or loading**, rather than the training process.为了进一步诊断问题，我建议先听**声码音频（直接从声码器输出 **），再进行完整的 TTS 推断。如果这个阶段音频已经听起来不自然或失真，问题很可能出在**声码器配置或加载**上，而不是训练过程。
> 
> Therefore, the robotic and unclear audio you are experiencing is not necessarily caused by the absence of a dedicated 16k vocoder. It is more likely due to a **configuration mismatch or incorrect vocoder setup**.因此，你听到的机械化且不清晰的音频，并不一定是因为没有专用的 16k 声码器。更可能是**配置不匹配或声码器设置不正确 **。
> 
> Thank you for your time, and please let me know if you need further clarification.感谢您的时间，如果需要进一步说明，请告诉我。
/home/hanzhiying/emospherepptest/EmoSpherepp/egs/egs_bases/tts/base.yaml：# task
base_config:
  - ../config_base.yaml
  - ./dataset_params.yaml
for_extractor: true


#############
# dataset in training
#############
endless_ds: true
min_frames: 0
max_frames: 1548
frames_multiple: 1
max_input_tokens: 1550
ds_workers: 1


###########
# optimization
###########
lr: 0.0005
scheduler: warmup # rsqrt|warmup|none
warmup_updates: 4000
optimizer_adam_beta1: 0.9
optimizer_adam_beta2: 0.98
weight_decay: 0
clip_grad_norm: 1
clip_grad_value: 0




###########
# train and eval
###########
use_word_input: false
max_valid_sentences: 1
max_valid_tokens: 60000
valid_infer_interval: 10000
train_set_name: 'train'
train_sets: ''
valid_set_name: 'valid'
test_set_name: 'test'
num_valid_plots: 10
test_ids: [ ]
test_input_yaml: ''
vocoder: HifiGan
vocoder_ckpt: '/home/hanzhiying/emospherepptest/EmoSpherepp/pretrained/checkpoints/vocoder/hifigan_16k/model_ckpt_best.pt' #'/workspace/choddeok/hd0/checkpoints/vocoder/bigV_16k'
vocoder_cls: models.vocoder.hifigan.HiFiGAN
vocoder_config: /home/hanzhiying/emospherepptest/EmoSpherepp/configs/models/vocoder/hifigan_16k.yaml
profile_infer: false
out_wav_norm: false
save_gt: false
save_f0: false
gen_dir_name: ''


save_mel_npy: true
save_plot_png: false
valid_plots: [0, 1, 20, 21, 40, 41, 60, 61, 80, 81]

Hello, during training, the tr/dur_loss continues to decrease, but val/dur_loss and val/prior_loss first drop and then rise sharply, while val/total_loss decreases slowly or plateaus, and lr_0 remains almost unchanged.I set a 4000-step linear warmup, after which the learning rate stays constant. Even at 1.1 million training steps, the learning rate is still basically unchanged.Could you please advise whether I should set the scheduler to rsqrt after 4000 steps and continue training up to 1.1 million steps?Thank you very much for your help and guidance.
Besides, I'm a second-year graduate student with poor coding skills, a total academic rookie. I'm begging for the visualization and evaluation code! Here's my email: 2093046865@qq.com. Please, kind expert, help a poor student out.
